### PR TITLE
First network

### PR DESCRIPTION
**Pull request by elizabethfeden**
* Created helpers.py with functions to be used in different notebooks
* Created notebook with unet basic training (no data preprocessing, default hyperparameters)

### CODE REVIEW

File: first-network.ipynb [hunk 1]
```diff
@@ -0,0 +1 @@
...et(X_train, y_train)\n","test = Dataset(X_test, y_test)"],"id":"oXsaAeoAYhjJ","execution_count":19,"outputs":[]},{"cell_type":"code","metadata":{"id":"IF2pWr81YtI9","executionInfo":{"status":"ok","timestamp":1635942768670,"user_tz":-180,"elapsed":4330,"user":{"displayName":"Elizabeth Fedenya","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjOHxCW3CMTv65QjqHm_9IP7FE7tXAORIyCLDEQ=s64","userId":"06503785506966883525"}}},"source":["model = Network(smp.Unet(in_channels=1, classes=4, activation='softmax'),\n","                nn.CrossEntropyLoss())"],"id":"IF2pWr81YtI9","execution_count":20,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Zxz8JXPsYvTQ","executionInfo":{"status":"ok","timestamp":1635943260624,"user_tz":-180,"elapsed":491971,"user":{"displayName":"Elizabeth Fedenya","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjOHxCW3CMTv65QjqHm_9IP7FE7tXAORIyCLDEQ=s64","userId":"06503785506966883525"}},"outputId":"eb8c1f73-c4b4-45bb-925e-f2107e2c67a6"},"source":["model.fit(train, epochs=20, verbose=True)"],"id":"Zxz8JXPsYvTQ","execution_count":21,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/py
```
Comment (Vitonka) [inline first-network.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #9.          size = len(data)
</pre><p>Нигде дальше не используется</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

Comment (Vitonka) [inline first-network.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #12.            for batch, (X, y) in enumerate(data):
</pre><p>Зачем делать enumerate, если затем batch нигде не используется?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

Comment (Vitonka) [inline first-network.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #24.        def fit_best(self, train, epochs=50, test_metric='accuracy'):
</pre><p>А я правильно понимаю, что в этой функции 50 раз обучается модель и из них выбирается лучшая?</p><p>UPD: а, нет, тут что-то более странное происходит. Поясните, чего тут хотелось добиться :)</p><p>UPD2: о, так это ещё и не используется нигде, если я правильно понимаю</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (Vitonka) [inline first-network.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #26.          lengths = [int(size * 0.8), size - int(size * 0.8)]
</pre><p>Магическая константа 0.8, нужно бы её как-то объявить</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

Comment (Vitonka) [inline first-network.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #33.            if i % 3 == 0:
</pre><p>Магическая константа 3</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

Comment (Vitonka) [inline first-network.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #65.                iou = self._iou(pred, y, args[0])
</pre><p>Лишнее</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

Comment (Vitonka) [inline first-network.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #66.                total_metric += self._iou(pred, y, args[0])
</pre><p>Что-то мне не очень нравится полагаться на args. Я тут как минимум жду kwargs с assert'ом на наличие определённого ключа, если выбрана метрика iou. А как максимум даже и говорить не буду, чего жду, пока что и с kwargs сойдёт :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (Vitonka) [inline first-network.ipynb]: <p>График бы!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='15'/>

Comment (Vitonka) [inline first-network.ipynb]: <p>Подписи к столбцам и строкам бы!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (Vitonka) [inline first-network.ipynb]: <p>В целом ничего удивительного: задача отделения фона для таких данных довольно простая, а UNet довольно хорошая сетка</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='19'/>

Comment (Vitonka) [inline first-network.ipynb]: <p>Тут бы (да и везде в выводе метрик) подписать про что они, для каких классов и так далее.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (Vitonka) [inline first-network.ipynb]: <p>И тут подписей не хватает!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/elizabethfeden/covid-tomography/pull/2/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='27'/>

File: helpers.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from PIL import Image
+from typing import List, Tuple, Optional
```
Comment (Vitonka) [inline helpers.py:6]: О, прям с указанием типов, моё уважение!

File: helpers.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    datasets.append(d)
+    print(f'Added {filename}, shape {d.shape}')
+  return datasets
+
+
+def load_datasets(prefix_path: str,
+                  size: Optional[Tuple[int, int]] = None
+                  ) -> List[np.ndarray]:
+  return load_datasets_by_names(
+      [prefix_path + name for name in DATASET_NAMES],
```
Comment (Vitonka) [inline helpers.py]: Тут лучше вот как поступить. Вообще использовать + для конкатенации путей - такое себе. В крайнем случае можно юзать os.path.join.
Но обычно все пользуются библиотекой под названием pathlib. Использование примерно такое было бы в вашем случае:
```
from pathlib import Path

# ниже строчка это из ноутбука, убираем бэкслеш в конце пути
path = Path('drive/MyDrive/kursach')

DATASET_NAMES = [
      Path('raw-data-kaggle/images_medseg.npy'),
      Path('raw-data-kaggle/masks_medseg.npy'),
      Path('raw-data-kaggle/images_radiopedia.npy'),
      Path('raw-data-kaggle/masks_radiopedia.npy'),
]

# и из этой строчки к которой коммент
[prefix_path / name for name in DATASET_NAMES],
```

**Review by Vitonka — DISMISSED** [review]

File: first-network.ipynb (hunk 1 above)
Comment (elizabethfeden) [reply first-network.ipynb]: <p>Изначальная идея была в том, чтобы разбить каждую эпоху на трейн-фазу и тест-фазу. Каждые n эпох тренировочная и тестовая выборки перетасовываются. Обучение останавливается, когда метрика, по которой мы тестируем, перестает расти. Таким образом мы как будто боремся с переобучением (наверное). По крайней мере, где-то видела такую штуку, но по факту не знаю, насколько это хорошая практика.</p><p> </p><p>Нигде не используется, потому что весь этот процесс работает сильно дольше, чем просто fit(), и пока в нем не было сильной необходимости, потому что я хотела посмотреть, как базово будет вести себя модель, а не обучить ее с наилучшими возможными параметрами. И еще в данной ситуации мне было не совсем понятно, по какой именно метрике тестировать в тестовой фазе.</p>

Comment (elizabethfeden) [reply first-network.ipynb]: <p>Давайте как максимум, чего уж там)</p><p> </p>

Comment (elizabethfeden) [reply first-network.ipynb]: <p>Сделано</p>

Comment (elizabethfeden) [reply first-network.ipynb]: <p>Сделала к столбцам, к строкам не совсем понимаю что подписывать</p>

Comment (elizabethfeden) [reply first-network.ipynb]: <p>Здесь сделала, в остальных местах вроде понятно по коду</p>

Comment (elizabethfeden) [reply first-network.ipynb]: <p>Сделано</p>
